diff --git a/aboutdata.h b/aboutdata.h
old mode 100644
new mode 100755
index 90cfa9532..e87a05c3b
--- a/aboutdata.h
+++ b/aboutdata.h
@@ -16,9 +16,9 @@
 inline KAboutData okularAboutData()
 {
     KAboutData about(QStringLiteral("okular"),
-                     i18n("Okular"),
+                     i18n("Okular-MMB"),
                      QStringLiteral(OKULAR_VERSION_STRING),
-                     i18n("Okular, a universal document viewer"),
+                     i18n("Okular-MMB, a universal document viewer with middle-click autoscroll"),
                      KAboutLicense::GPL,
                      i18n("(C) 2002 Wilco Greven, Christophe Devriese\n"
                           "(C) 2004-2005 Enrico Ros\n"
diff --git a/part/pageview.cpp b/part/pageview.cpp
old mode 100644
new mode 100755
index 93e9567ed..d859d1427
--- a/part/pageview.cpp
+++ b/part/pageview.cpp
@@ -227,6 +227,12 @@ public:
     QPoint dragScrollVector;
     QTimer dragScrollTimer;
 
+    // middle click autoscroll
+    bool scrollWithMiddleMouseButton = false;
+    QTimer *middleClickAutoscrollTimer = nullptr;
+    QPoint middleClickAutoscrollOrigin;
+    bool middleClickAutoscrollWasDragged = false;
+
     // left click depress
     QTimer leftClickTimer;
 
@@ -2282,24 +2288,11 @@ void PageView::mouseMoveEvent(QMouseEvent *e)
         return;
     }
 
-    // if holding mouse mid button, perform zoom
-    if (e->buttons() & Qt::MiddleButton) {
-        int deltaY = d->mouseMidLastY - e->globalPosition().y();
-        d->mouseMidLastY = e->globalPosition().y();
-
-        const float upperZoomLimit = d->document->supportsTiles() ? 99.99 : 3.99;
-
-        // Wrap mouse cursor
-        if (Okular::Settings::dragBeyondScreenEdges()) {
-            Qt::Edges wrapEdges;
-            wrapEdges.setFlag(Qt::TopEdge, d->zoomFactor < upperZoomLimit);
-            wrapEdges.setFlag(Qt::BottomEdge, d->zoomFactor > 0.101);
-
-            deltaY += CursorWrapHelper::wrapCursor(e->globalPosition().toPoint(), wrapEdges).y();
+    // if in autoscroll mode, handle dragging
+    if (d->scrollWithMiddleMouseButton) {
+        if ((e->globalPosition().toPoint() - d->middleClickAutoscrollOrigin).manhattanLength() > QApplication::startDragDistance()) {
+            d->middleClickAutoscrollWasDragged = true;
         }
-
-        // update zoom level, perform zoom and redraw
-        continuousZoom(deltaY);
         return;
     }
 
@@ -2421,11 +2414,25 @@ void PageView::mousePressEvent(QMouseEvent *e)
         d->autoScrollTimer->stop();
     }
 
-    // if pressing mid mouse button while not doing other things, begin 'continuous zoom' mode
+    // check if we are already autoscrolling, if so stop it and consume click
+    if (d->scrollWithMiddleMouseButton) {
+        d->scrollWithMiddleMouseButton = false;
+        d->middleClickAutoscrollTimer->stop();
+        updateCursor();
+        return;
+    }
+
+    // if pressing mid mouse button while not doing other things, begin 'autoscroll' mode
     if (e->button() == Qt::MiddleButton) {
-        d->mouseMidLastY = e->globalPosition().y();
-        setCursor(Qt::SizeVerCursor);
-        CursorWrapHelper::startDrag();
+        d->scrollWithMiddleMouseButton = true;
+        d->middleClickAutoscrollWasDragged = false;
+        d->middleClickAutoscrollOrigin = e->globalPosition().toPoint();
+        setCursor(Qt::SizeAllCursor);
+        if (!d->middleClickAutoscrollTimer) {
+            d->middleClickAutoscrollTimer = new QTimer(this);
+            connect(d->middleClickAutoscrollTimer, &QTimer::timeout, this, &PageView::slotMiddleClickAutoscroll);
+        }
+        d->middleClickAutoscrollTimer->start(1000 / 60); // 60 fps
         return;
     }
 
@@ -2665,9 +2672,16 @@ void PageView::mouseReleaseEvent(QMouseEvent *e)
 
     const QPoint eventPos = contentAreaPoint(e->pos());
 
-    // handle mode independent mid bottom zoom
+    // handle mode independent mid button release
     if (e->button() == Qt::MiddleButton) {
-        continuousZoomEnd();
+        if (d->scrollWithMiddleMouseButton) {
+            // if we dragged, stop autoscroll
+            if (d->middleClickAutoscrollWasDragged) {
+                d->scrollWithMiddleMouseButton = false;
+                d->middleClickAutoscrollTimer->stop();
+                updateCursor();
+            }
+        }
         return;
     }
 
@@ -5145,6 +5159,30 @@ void PageView::slotDragScroll()
     updateSelection(p);
 }
 
+void PageView::slotMiddleClickAutoscroll()
+{
+    if (!d->scrollWithMiddleMouseButton) return;
+
+    QPoint delta = QCursor::pos() - d->middleClickAutoscrollOrigin;
+
+    // add some deadzone and scaling factor
+    int dx = 0;
+    int dy = 0;
+    const int deadzone = 10;
+    const int scale = 10;
+
+    if (abs(delta.x()) > deadzone) {
+        dx = (delta.x() > 0 ? delta.x() - deadzone : delta.x() + deadzone) / scale;
+    }
+    if (abs(delta.y()) > deadzone) {
+        dy = (delta.y() > 0 ? delta.y() - deadzone : delta.y() + deadzone) / scale;
+    }
+
+    if (dx != 0 || dy != 0) {
+        scrollTo(horizontalScrollBar()->value() + dx, verticalScrollBar()->value() + dy);
+    }
+}
+
 void PageView::slotShowWelcome()
 {
     // show initial welcome text
diff --git a/part/pageview.h b/part/pageview.h
old mode 100644
new mode 100755
index 92552c451..cd98d09d7
--- a/part/pageview.h
+++ b/part/pageview.h
@@ -294,6 +294,8 @@ private Q_SLOTS:
     void slotAutoScroll();
     // activated by the dragScroll timer
     void slotDragScroll();
+    // activated by the middleClickAutoscroll timer
+    void slotMiddleClickAutoscroll();
     // show the welcome message
     void slotShowWelcome();
     // activated by left click timer
